packages/babel-core/src/config/caching.js

```diff
@@ -14,91 +27,253 @@ type SimpleCacheConfiguratorObj = {
+
+    const handlerResult = handler(arg, cache);
+
+    const gen = isIterableIterator(handlerResult);
+
+    // $FlowIgnore
+    const value: ResultT = gen ? yield* handlerResult : handlerResult;
+
+    updateFunctionCache(callCache, cache, arg, value);
+
+    if (asyncContext) {
+      /*:: invariant(finishLock) */
+      futureCache.delete(arg);
+      finishLock.release(value);
+    }
+
+    return value;
+  };
 }
 
 type CacheMap<ArgT, ResultT, SideChannel> =
   | Map<ArgT, CacheEntry<ResultT, SideChannel>>
   | WeakMap<ArgT, CacheEntry<ResultT, SideChannel>>;
 
-function makeCachedFunction<
+function* getCachedValue<
   ArgT,
   ResultT,
   SideChannel,
   // $FlowIssue https://github.com/facebook/flow/issues/4528
   Cache: CacheMap<ArgT, ResultT, SideChannel>,
 >(
-  callCache: Cache,
-  handler: (ArgT, CacheConfigurator<SideChannel>) => ResultT,
-): (ArgT, SideChannel) => ResultT {
-  return function cachedFunction(arg, data) {
-    let cachedValue: CacheEntry<ResultT, SideChannel> | void = callCache.get(
-      arg,
-    );
-
-    if (cachedValue) {
-      for (const { value, valid } of cachedValue) {
-        if (valid(data)) return value;
-      }
+  cache: Cache,
+  arg: ArgT,
+  data: SideChannel,
+): Handler<{ valid: true, value: ResultT } | { valid: false, value: null }> {
+  const cachedValue: CacheEntry<ResultT, SideChannel> | void = cache.get(arg);
+
+  if (cachedValue) {
+    for (const { value, valid } of cachedValue) {
+      if (yield* valid(data)) return { valid: true, value };
     }
+  }
 
-    const cache = new CacheConfigurator(data);
+  return { valid: false, value: null };
+}
+
+/* NOTE: This comment refers to both the getCachedValueOrWait and setupAsyncLocks functions.
+ *
+ * > There are only two hard things in Computer Science: cache invalidation and naming things.
+ * > -- Phil Karlton
+ *
+ * I don't know if Phil was also thinking about handling a cache whose invalidation function is
+ * defined asynchronously is considered, but it is REALLY hard to do correctly.
+ *
+ * The implemented logic (only when gensync is run asynchronously) is the following:
+ *   1. If there is a valid cache associated to the current "arg" parameter,
+ *       a. RETURN the cached value
+ *   2. If there is a ConfigLock associated to the current "arg" parameter,
+ *       a. Wait for that lock to be released
+ *       b. GOTO step 1
+ *   3. If there is a FinishLock associated to the current "arg" parameter representing a valid cache,
+ *       a. Wait for that lock to be released
+ *       b. RETURN the value associated with that lock
+ *   4. Acquire to locks: ConfigLock and FinishLock
+ *   5. Start executing the cached function
+ *   6. When the cache if configured, release ConfigLock
+ *   7. Wait for the function to finish executing
+ *   8. Release FinishLock
+ *   9. Send the function result to anyone waiting on FinishLock
+ *   10. Store the result in the cache
+ *   11. RETURN the result
```

Comment:
@loganfsmyth What do you think abuot this algorithm, to solve the concurrency issue?
It is sub optimal in cases like the following one (it will take 2s instead of 1s), but I don't think that they are common:

```js
const fn = makeStrongCache(function* (arg) {
  yield* wait(1000);
  cache.never();
}).async;

fn(); fn();
```

I initially tried to force `cache.*` to be called synchronously (before any `yield`), but then found out that it doesn't even work for our internal code: https://github.com/babel/babel/pull/10507/files#diff-83f25d98662c0f8e8ad7cc7d1ed06f8eR32

Reply:
I tested the performance of this change, and there is a problem.
I have compiled every src file of this repo four times (405 x 4 files) and there are the results:

|           | **`cache.forever()`** | **`cache.never()`** |
|----------:|:---------------------:|:-------------------:|
| **async** |     15.372567052 s    |   470.434076412 s   |
|  **sync** |     15.862861386 s    |    19.743430078 s   |

As you can see, there is a *big* problem with `cache.never()` + async.
It is related to re-compiling the same file. If I compile every file once instead of four times, it doesn't have this big difference:

|           | **`cache.never()`** |
|----------:|:-------------------:|
| **async** |    10.391877456 s   |
|  **sync** |    6.149023702 s    |

Reply:
@loganfsmyth We can solve this big performance problem if we disallow the cache to be configured asynchronously. Synchronously configuring the cache with an async validator would still be allowed:

```js
// waitFor transforms a "Promise<T> | T" into a "Gensync<T>"

makeWeakCache(function* (arg) {
  // Allowed, because cache.using is called synchronously

  const val = yield* waitFor(
    cache.using(async () => {
      await wait(100);
      return process.env.FOO;
    })
  );
});


makeWeakCache(function* (arg) {
  // Disallowed, because cache.using is called asynchronously

  yield* waitFor(wait(100));
  const val = cache.using(() => process.env.FOO);
});
```

In the future, this change will be observable to our users because they won't be able to use `await` before configuring the cache.

In order to implement this behaviour I had to modify `gensync`'s internals, because I couldn't find a way "from the outside" of knowing if a generator has been suspended. I will open a PR soon because I edited it in my `node_modules` and I fear to loose those changes, but if you have a better solution feel free to propose it!

Reply:
These are the new numbers:

| **#** |    **async**    | **sync**       | **sync `master`** |
|:-----:|:---------------:|:--------------:|:-----------------:|
|  100  |  2.219376366 s  |  2.256803449 s | 2.143184428 s     |
|  200  |  3.714403922 s  |  3.641993385 s | 3.591130842 s     |
|  400  |  6.338499966 s  |  6.638144645 s | 6.299714871 s     |
|  800  |  11.772867713 s | 12.198001927 s | 11.539516664 s    |
|  1600 |  21.811404623 s | 22.632490371 s | 21.436522896 s    |